### PR TITLE
CP detail view endpoint

### PR DESCRIPTION
**Pull request by juliasaveliff**

### CODE REVIEW

File: rackcity/utils/change_planner_utils.py [hunk 1]
```diff
@@ -178,3 +178,117 @@ def get_page_count_response_for_cp(request, change_plan):
...
     page_count = math.ceil(count / page_size)
     return JsonResponse({"page_count": page_count})
+
+
+def get_changes_on_asset(asset, assetCP):
+    fields = [field.name for field in Asset._meta.fields]
+    changes = []
+    for field in fields:
+        if getattr(asset, field) != getattr(assetCP, field):
+            # TODO will need to serialize field if it's another model
```
Comment (clairecfu) [inline rackcity/utils/change_planner_utils.py]: so this would be rack/model... what about network and power connections? since those aren't fields on asset. You could you reverse relations to get all the network connections and power connections that Asset and AssetCP are forieign keys for and compare those? if you want to leave that in a separate PR im also good with that.

Comment (clairecfu) [reply rackcity/utils/change_planner_utils.py]: but we would need to do that in order to highlight changes to network /power connections

File: rackcity/utils/change_planner_utils.py [hunk 2]
```diff
@@ -178,3 +178,117 @@ def get_page_count_response_for_cp(request, change_plan):
...
+            })
+    return changes
+
+
+def get_cp_modification_conflicts(asset_cp):
+    conflicts = []
+    nonresolvable_message = "This conflict cannot be resolved; the change " + \
+        "needs to be removed from your change plan."
+    conflicting_asset_message_1 = "Due to more recent live changes, " + \
+        "your change planner modification to this asset's "
```
Comment (clairecfu) [inline rackcity/utils/change_planner_utils.py]: small nit but maybe instead of modifications, say `your change plan's version of this's assets: ` because a conflict could come from a created asset on cp.

File: rackcity/utils/change_planner_utils.py [hunk 3]
```diff
@@ -178,3 +178,117 @@ def get_page_count_response_for_cp(request, change_plan):
...
+def get_modifications_in_cp(change_plan):
+    assets_cp = AssetCP.objects.filter(change_plan=change_plan)
+    modifications = []
+    for asset_cp in assets_cp:
+        assetCP_data = RecursiveAssetCPSerializer(asset_cp).data
+        related_asset = asset_cp.related_asset
+        if asset_cp.is_decommissioned:
+            title = "Decommission asset"
+            if asset_cp.asset_number:
+                title += " " + str(asset_cp.asset_number)
```
Comment (clairecfu) [inline rackcity/utils/change_planner_utils.py]: if the asset_cp does not have an asset number, this is going to say Decomission asset None. Can we add a check for this and either leave it out or use its rack + rack location if the asset number doesn't exist?

Comment (clairecfu) [reply rackcity/utils/change_planner_utils.py]: "Decomission/Modify/Create" asset [asset_number] on [Rack] at [location]. This would be a helpful string if we use this for work orders as well. But add logic so that if asset_number is none it is just omitted and doesn't print as None

File: rackcity/utils/change_planner_utils.py [hunk 4]
```diff
@@ -178,3 +178,117 @@ def get_page_count_response_for_cp(request, change_plan):
...
+        else:
+            title = "Create asset"
+            if asset_cp.asset_number:
+                title += " " + str(asset_cp.asset_number)
+            asset_data = None
+        modifications.append({
+            "title": title,
+            "asset": asset_data,
+            "asset_cp": assetCP_data,
+            "conflicts": get_cp_modification_conflicts(asset_cp)
```
Comment (clairecfu) [inline rackcity/utils/change_planner_utils.py]: can you also include `change typ` and `changes` field? as documented on the notion

File: rackcity/utils/change_planner_utils.py (hunk 2 above)
Comment (juliasaveliff) [reply rackcity/utils/change_planner_utils.py]: cool I like

File: rackcity/utils/change_planner_utils.py (hunk 3 above)
Comment (juliasaveliff) [reply rackcity/utils/change_planner_utils.py]: I'm already checking `if asset_cp.asset_number:` here! But I like adding more detail about location I'll make that change

File: rackcity/utils/change_planner_utils.py [hunk 5]
```diff
@@ -178,3 +187,133 @@ def get_page_count_response_for_cp(request, change_plan):
...
     return JsonResponse({"page_count": page_count})
+
+
+def get_changes_on_asset(asset, assetCP):
+    fields = [field.name for field in Asset._meta.fields]
+    changes = []
+    for field in fields:
+        if getattr(asset, field) != getattr(assetCP, field):
+            changes.append(field)
+    # TODO check if network & power connections have changed
```
Comment (juliasaveliff) [inline rackcity/utils/change_planner_utils.py]: @clairecfu will do this in another PR

File: rackcity/utils/change_planner_utils.py [hunk 6]
```diff
@@ -178,3 +187,133 @@ def get_page_count_response_for_cp(request, change_plan):
...
+                title += get_location_detail(asset_cp)
+            asset_data = None
+        modifications.append({
+            "title": title,
+            "type": modification_type.value,
+            "asset": asset_data,
+            "asset_cp": assetCP_data,
+            "conflicts": get_cp_modification_conflicts(asset_cp),
+            "changes": changes,
+        })
```
Comment (juliasaveliff) [inline rackcity/utils/change_planner_utils.py]: @clairecfu updated the API object on notion slightly, check dat out

File: rackcity/utils/change_planner_utils.py [hunk 7]
```diff
@@ -178,3 +187,133 @@ def get_page_count_response_for_cp(request, change_plan):
...
+        if asset_cp.is_conflict:
+            conflicts.append({
+                "conflict_message":
+                    "Live changes have been made to this asset since your " +
+                    "latest change planner modification. This conflict " +
+                    "needs to be resolved. Please select which version " +
+                    "you would like to keep.",
+                "conflicting_asset": None,
+                "conflict_resolvable": True,
+            })
```
Comment (clairecfu) [inline rackcity/utils/change_planner_utils.py]: based on my comments on the slack thread, `if asset_cp.is_conflict && asset_cp.related_asset is None`: to check if the asset has been deleted on msater. and then mark this as a unresolvable conflict? conflict message `The asset being modified on this change plan has been deleted live. + non resolvable message`

File: rackcity/utils/change_planner_utils.py [hunk 8]
```diff
@@ -178,3 +187,138 @@ def get_page_count_response_for_cp(request, change_plan):
...
+    nonresolvable_message = "This conflict cannot be resolved; the " + \
+        "related changes need to be removed from your change plan."
+    conflicting_asset_message_1 = "Due to more recent live changes, " + \
+        "your change plan version of this asset's "
+    conflicting_asset_message_2 = " now conflicts with a live asset. "
+    if asset_cp.is_conflict and asset_cp.related_asset is None:
+        conflicts.append({
+            "conflict_message":
+                "This asset has been deleted live. " + nonresolvable_message
+        })
```
Comment (juliasaveliff) [inline rackcity/utils/change_planner_utils.py]: @clairecfu added here
